Add "Replace question" action to dashboard cards

kulyk opened the pull request:
Closes #36497, closes #36586

Adds a "Replace" button to dashboard cards with questions or models. It makes it possible to swap the dashboard card's query while keeping the dashboard layout intact. It's important to call out that we would reset properties of a dash card that no longer makes sense with a new query: parameter mappings, click behavior, visualization settings, and added series.

### How to verify

1. Open any dashboard that has a few cards
2. Enter the editing mode
3. Hover a dashboard card using a question or a model
4. Find and click the "Replace" action in the top-right floating panel
5. Select a new question or a model from the modal picker
6. Ensure the dashboard card's content is replaced, and the query is rerun
7. Ensure parameters auto-wiring worked (your dashboard should have another card using the same table and mapped to some parameters)
8. Ensure you can undo the revert
9. Ensure changes are persisted after clicking "Save"

### Demo

<details>
<summary>Basic</summary>

https://github.com/metabase/metabase/assets/17258145/15363bc3-7a37-49f4-90b6-b3af2096c842

</details>

<details>
<summary>Filters auto-wiring</summary>

https://github.com/metabase/metabase/assets/17258145/38d75e0f-2185-470a-926f-bca7ab74c498

</details>

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

kulyk committed:
Remove `isVirtualDashCard` prop

kulyk committed:
Add "replace" button to `DashCardActionsPanel`

kulyk committed:
Add `QuestionPickerModal`

kulyk committed:
Add `replaceCard` action

kulyk committed:
Use `/card/:id/query` endpoint for swapped cards

kulyk committed:
Implement card replace

kulyk committed:
Add basic undo toast

markbastian committed:
`update-dashboard-card` will update `:card_id`

Previously, `card_id` in `update-dashboard-card!` only applied if the dashcard was an action card. There was a cryptic comment about this, but it also indicates that maybe it's ok to update.

This unsticks FE work but still needs BE unit tests.

markbastian committed:
Adding unit test for updating dashcard's card

The new test checks the state of a dashboard with a GET then PUTs the update setting the new card id and verifies the result.

kulyk committed:
Clean up `DashCard` test suite

kulyk committed:
Add tests for `DashCard` component

kulyk committed:
Add tests for `QuestionPickerModal` component

kulyk committed:
Add tests for `replaceCard` action

kulyk committed:
Remove redundant action argument

kulyk committed:
Add test for parameters auto-wiring

kulyk committed:
Fix unexpected card query endpoint usage

kulyk committed:
Use more accessible selector in `saveDashboard`

kulyk committed:
Add e2e tests

kulyk committed:
Remove `justAdded` from `createMockDashboardCard`

kulyk committed:
Fix test setup

cdeweyx reviewed (approved):
Working like a charm 👍

markbastian committed:
Updating `update-action-cards-test` test

Under the old behavior, only action dashcards allowed changing their card ids. Now card ids can be changed for any dashcard type. This test captured the old behavior and has been updated to capture the new behavior.

kulyk committed:
Remove previous events

kulyk committed:
Track `dashboard_card_replaced` event

markbastian committed:
Merge remote-tracking branch 'origin/36497-replace-dashcard-card' into 36497-replace-dashcard-card

snoe reviewed (approved):
BE LGTM

calherries commented:
If Case is cool with updating card IDs, you should change the logic in `models.dashboard-card.update-dashboard-card!` which prevents it https://github.com/metabase/metabase/blob/ffe0385b01b698b02d66b9d18b4dcc238b200d7d/src/metabase/models/dashboard_card.clj#L182-L184

markbastian commented:
> If Case is cool with updating card IDs, you should change the logic in `models.dashboard-card.update-dashboard-card!` which prevents it
> 
> https://github.com/metabase/metabase/blob/ffe0385b01b698b02d66b9d18b4dcc238b200d7d/src/metabase/models/dashboard_card.clj#L182-L184

Isn't this addressed here? https://github.com/metabase/metabase/pull/36744/files#diff-b6abb82aa02f4647f6ac74d17bfa75dda523bf75c2eed11954c56127a6d7446dR180-R189

alxnddr commented:
It would be great if users could replace errored cards, I think it can be a valid use case for the functionality
<img width="870" alt="Screenshot 2023-12-13 at 6 43 34 PM" src="https://github.com/metabase/metabase/assets/14301985/02af0592-0bd2-4340-87b1-3a87a6c6aa27">

alxnddr commented on frontend/src/metabase/dashboard/selectors/selectors-typed.ts:
```diff
@@ -5,6 +5,10 @@ export function getDashboardId(state: State) {
 }
 
+export function getDashboardBeforeEditing(state: State) {
+  return state.dashboard.isEditing;
```
Oh, `isEditing` is a dashboard instead of a boolean 🤯

kulyk replied:
I'm afraid you're right 😢

alxnddr reviewed (approved):
Nice and clean! Please have a look at the suggestion about errored cards

kulyk committed:
Merge branch 'master' into 36497-replace-dashcard-card

kulyk committed:
Show "replace" action for erroring queries

kulyk commented:
> It would be great if users could replace errored cards, I think it can be a valid use case for the functionality

@alxnddr, love the idea, added in [`92bc305`](https://github.com/metabase/metabase/pull/36744/commits/92bc305d81eacc53e01acbdfb9b15c4ebc2448d8). Thanks!

https://github.com/metabase/metabase/assets/17258145/1b1085fd-99ba-4752-bcd6-5b4df064aa24

perivamsi commented:
> > It would be great if users could replace errored cards, I think it can be a valid use case for the functionality
> 
> @alxnddr, love the idea, added in [`92bc305`](https://github.com/metabase/metabase/pull/36744/commits/92bc305d81eacc53e01acbdfb9b15c4ebc2448d8). Thanks!
> 
> Demo

If it is possible to show the "replace card" prompt directly on the error card, that will be helpful.

So change "There was a problem displaying this card." by adding a "replace it" link or button at the end? This is more discoverable.

kulyk commented:
> > > It would be great if users could replace errored cards, I think it can be a valid use case for the functionality
> 
> > 
> 
> > @alxnddr, love the idea, added in [`92bc305`](https://github.com/metabase/metabase/pull/36744/commits/92bc305d81eacc53e01acbdfb9b15c4ebc2448d8). Thanks!
> 
> > 
> 
> > Demo
> 
> 
> 
> If it is possible to show the "replace card" prompt directly on the error card, that will be helpful.
> 
> 
> 
> So change "There was a problem displaying this card." by adding a "replace it" link or button at the end? This is more discoverable.

For context, we figured we'd prefer to keep the action as is. People might prefer to fix the issue instead of replacing the card
